### PR TITLE
Home Tag Filtering & Pagination

### PR DESCRIPTION
**Pull request by delusioninabox**
I saw there was a PR for pagination, but it hasn't been updated in some time... I also wanted to give it a go on my own to practice with Ember. 💦 This PR has the full functionality for the home list of articles, toggling through tags, pagination, and includes tests. (See checklist below for specific tasks done.)

**On the home route as a user, I can:**
- [x]   change a tag to browse
- [x]   change the page to browse

**Technical specifications:**
- [x]   page count matches the total article count
- [x]   page count updates when a tag is changed
- [x]   current page is highlighted
- [x]   10 posts per page
- [x]   passes linters
- [x]   set post limit as easily changed variable
- [x]   pagination is a component

**Tests:**
- [x]   visiting homepage
- [x]   clicking a tag
- [x]   clicking a page
- [x]   resetting tag/page

### CODE REVIEW

**Comment by delusioninabox** [comment]
Encountered a weird bug where tests pass in the browser but not phantomjs in the browser. Resolved thanks to the help of this (https://github.com/ember-cli/ember-cli/issues/7001), but we should probably upgrade the Ember version next. 😅

**Comment by Alonski** [comment]
Hey @delusioninabox,
First of all thanks for the awesome work!
I missed this PR but this looks pretty great.
I would love to take a look at this.
Would it be possible to fix conflicts before I look at this?

**Comment by delusioninabox** [comment]
After rebasing and upgrading, I keep getting this error on my tests:
`TypeError: Cannot read property 'createList' of undefined`
(For any time I use `server.createList('model')`)

Any chance you know why that keeps happening? 😩 I've checked that server is in the es file, so I'm a bit boggled.

**Comment by jelhan** [comment]
> After rebasing and upgrading, I keep getting this error on my tests:
> TypeError: Cannot read property 'createList' of undefined
> (For any time I use server.createList('model'))

The tests in `tests/acceptance/home-test.js` are not inside of `module`, in which `setupMirage` is called. Therefore mirage is not booted up and `server` is `undefined`. Just move them inside the callback of `module` and everything should be fine.

**Comment by delusioninabox** [comment]
@jelhan Thanks for catching that!!! 🙏

**Comment by delusioninabox** [comment]
@Alonski Ready for review! 🙌

**Comment by Alonski** [comment]
Hey, I will try to take a look at this in the coming days. 
Thanks!

File: .eslintrc.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module.exports = {
   globals: {
-    server: true
+    server: true,
```
Comment (Alonski) [inline .eslintrc.js]: Weird that ESLint isn't catching this but that comma shouldn't be there 😅

File: .vscode/settings.json [hunk 2]
```diff
@@ -9,7 +9,6 @@
     "**/assets": true,
     "**/dist": true
   },
-  "editor.formatOnSave": true,
```
Comment (Alonski) [inline .vscode/settings.json]: Was this annoying you while working?

File: app/routes/home.js [hunk 3]
```diff
@@ -2,10 +2,31 @@ import Route from '@ember/routing/route';
...
+        limit: 10,
+        offset: (params.page - 1) * perPage
+      })
     });
+  },
+
+  setupController(controller, model) {
+    this._super(controller, model);
+    controller.set('totalPages', model.articles.get('meta').articlesCount);
+    controller.set('postsPerPage', model.perPage);
```
Comment (Alonski) [inline app/routes/home.js]: Is it possible to refactor this into the controller?
Something like: postsPerPage: oneWay/readOnly/alias(model.perPage)?

File: app/serializers/article.js [hunk 4]
```diff
@@ -17,6 +17,9 @@ export default DS.RESTSerializer.extend(DS.EmbeddedRecordsMixin, {
       });
       return newArticle;
     });
+    newPayload.meta = {
+      articlesCount: payload.articlesCount
+    };
```
Comment (Alonski) [inline app/serializers/article.js]: This seems to cause a warning:
Encountered "articlesCount" in payload, but no model was found for model name "articles-count" (resolved model name using (unknown).modelNameFromPayloadKey("articlesCount"))

Can you look into this please?

File: app/controllers/home.js [hunk 5]
```diff
@@ -4,6 +4,17 @@ import { oneWay } from '@ember/object/computed';
...
+  page: 1,
+  totalPages: 1,
 
-  isAuthenticated: oneWay('session.isAuthenticated')
+  isAuthenticated: oneWay('session.isAuthenticated'),
+
+  actions: {
+    setVariable(variable, value) {
+      if (this.get(variable) !== value && variable === 'tag') this.set('page', 1);
+      this.set(variable, value);
```
Comment (Alonski) [inline app/controllers/home.js]: This seems to only set the `tag` property. Any reason this isn't called setTag?

File: config/environment.js [hunk 6]
```diff
@@ -29,6 +29,9 @@ module.exports = function(environment) {
     'ember-simple-auth': {
       // Override the default `index` value
       routeIfAlreadyAuthenticated: 'home'
+    },
+    'ember-cli-mirage': {
+      enabled: true
```
Comment (Alonski) [inline config/environment.js]: Any reason this is needed?
The setup that we have is that if we run `yarn start` without params it works with Mirage.
If we run `yarn start:dev` it proxies to conduit api to run against real data.

This is causing Mirage to always work even when proxying.

File: ember-cli-build.js [hunk 7]
```diff
@@ -5,6 +5,9 @@ const EmberApp = require('ember-cli/lib/broccoli/ember-app');
 module.exports = function(defaults) {
   const app = new EmberApp(defaults, {
     // Add options here
+    babel: {
+      includePolyfill: true
+    }
```
Comment (Alonski) [inline ember-cli-build.js]: Which polyfill was needed for this?
Also this throws a deprecation. The correct syntax today is:
```
  'ember-cli-babel': {
    includePolyfill: true
  }
```

File: tests/.eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
+module.exports = {
```
Comment (Alonski) [inline tests/.eslintrc.js]: Any reason we need this file that we can do in the original top level .`eslintrc.js`?

File: mirage/config.js [hunk 9]
```diff
@@ -17,7 +17,20 @@ export default function() {
...
+    const allArticles = schema.articles.all(),
+      params = request.queryParams,
+      limit = parseInt(params.limit),
+      page = parseInt(params.offset) / limit,
+      start = page * limit,
+      end = start + limit,
+      newArticles = allArticles.models.slice(start, end),
+      newSchema = {};
+    newSchema.articles = newArticles;
+    newSchema.articlesCount = allArticles.length;
```
Comment (Alonski) [inline mirage/config.js]: Running in Mirage this seems to cause a warning:
`warn.js:45 WARNING: Encountered "articlesCount" in payload, but no model was found for model name "articles-count" (resolved model name using (unknown).modelNameFromPayloadKey("articlesCount"))`
It seems that articlesCount isn't being set into `meta` and Ember Data is instead trying to serialize this as a model

**Comment by Alonski** [comment]
Awesome work @delusioninabox!
Just a few fixes and we can get this merged

**Comment by delusioninabox** [comment]
Thanks for reviewing! I've got a lot going on atm, but I'll try to wrap this up this weekend. 🙌

**Comment by fire** [comment]
Ping!

**Comment by Alonski** [comment]
Hello @fire are you pinging this because you want to help out?

**Comment by fire** [comment]
I don't know the status of this, and I don't want to work on something someone else is already completing.

Spent some time trying the emberjs demo with phoenix elixir backend.

**Comment by fire** [comment]
I have a meeting scheduled later today, but I can join the discord afterwards.

What's your discord name?

**Comment by delusioninabox** [comment]
@fire @Alonski Sorry, my husband got a new job and we had to move, so last two months have been crazy. If you are able to finish the changes on this fire, that's be great!!! 🙏  I'm also on the Ember Discord if you need to ping me with any ?s.

**Comment by Alonski** [comment]
@fire My name on Discord is Alon.
@delusioninabox Family comes first :) If you ever find the time to help out some more that would be great whenever you can. Good luck!

**Comment by delusioninabox** [comment]
@Alonski Same as my GH, delusioninabox! 👍

**Comment by Alonski** [comment]
Made some changes here:
https://github.com/gothinkster/ember-realworld/tree/delusioninabox-feature/tag-filtering
@delusioninabox can you merge these into your PR?

**Comment by delusioninabox** [comment]
@Alonski Done! 👍

**Comment by Alonski** [comment]
After almost half a year this has been merged in :)

**Comment by delusioninabox** [comment]
@Alonski 😂🙈😅

**Pull request merged**